### ISSUE TITLE
cygwin: "Exception: Unsupported platform: cygwin" in xattr.py

### ISSUE THREAD

**Issue by jumper444** [jborg/attic#194]
I believe I have a functioning cygwin environment with all needed code installed.
I've been able to install Attic.  The only problem now is that file "xattr.py" has
some extended attribute config/settings for linux, darwin, and freebsd.  If that
file doesn't detect one of those OS strings, it fails with exception (see below).

I've looked at the xattr.py code and it is clearly defining and setting up
an environment depending on which OS it detects.  Here is my question...
isn't Cygwin essentially (at least) one of these other OS type environments?
I mean, yes...I know it's a Posix implementation on Windows, but doesn't the
implementation itself match (or closely match) one of these existing choices?
That would allow the detection code to be "linux" or "cygwin" for example.

If Cygwin doesn't match, then how hard is it to create this code section for?
And what about simply disabling "extended attribute" support (which I have
seen on other backup programs assuming this is the same type of thing).
The xattr.py code ALREADY has a test as the first function to determine:
"if xattr is enabled on the file system".
So what if we just default to no for cygwin environment.  Will that alternately work?

I'm not sure what xattr.py does but my assumption is that it deals with things
like symlinks, owners, and permissions that are not (similarly) present on NTFS, 
FAT, and the like.  It it handles none of those it is still partially valuable for backup.
And then perhaps a person could create appropriate such code for NTFS scheme
in the future.
```
$ attic
Traceback (most recent call last):
  File "/usr/bin/attic", line 2, in <module>
    from attic.archiver import main
  File "/usr/lib/python3.2/site-packages/attic/archiver.py", line 13, in <module>
    from attic.archive import Archive, ArchiveChecker
  File "/usr/lib/python3.2/site-packages/attic/archive.py", line 16, in <module>
    from attic import xattr
  File "/usr/lib/python3.2/site-packages/attic/xattr.py", line 251, in <module>
    raise Exception('Unsupported platform: %s' % sys.platform)
Exception: Unsupported platform: cygwin
```

**Comment by jumper444** [comment]
```
def is_enabled():
    """Determine if xattr is enabled on the filesystem
    """
    with tempfile.NamedTemporaryFile() as fd:
        try:
            setxattr(fd.fileno(), 'user.name', b'value')
        except OSError:
            return False
        return getxattr(fd.fileno(), 'user.name') == b'value'
```

This is defined for calling use outside of xattr.py code (nothing calls it inside module).
If it exists as such to tell a caller that attributes are not avail,
they why does the code below fail if it can't detect a known
OS?  Why not just return NOT enabled instead of failing on an unknown OS?

**Comment by jumper444** [comment]
Well...i hard coded around this module by setting `is_enabled()` to always return False
and then modded the lower down code to match "linux" AND "cygwin" (assuming a calling
program wouldn't call 'is_enabled' first.)  That change then results in helpers.py raising
ExtensionModuleError exception when `attic.platform.API_VERSION != 2' in check_extension_modules() function fails.  (attic.platform.API_VERSION is 1).

If I go to platform.py I find another OS type check (which returns that 1 if neither linux, darwin, bsd).
I tried also forcing this to 2, but got:
```
$ attic create -v ~/backup.attic::test1 /cygdrive/j/data
Initializing cache...
/cygdrive/j/quote.com
Traceback (most recent call last):
  File "/usr/bin/attic", line 3, in <module>
    main()
  File "/usr/lib/python3.2/site-packages/attic/archiver.py", line 727, in main
    exit_code = archiver.run(sys.argv[1:])
  File "/usr/lib/python3.2/site-packages/attic/archiver.py", line 717, in run
    return args.func(args)
  File "/usr/lib/python3.2/site-packages/attic/archiver.py", line 128, in do_create
    self._process(archive, cache, args.excludes, args.exclude_caches, skip_inodes, path, restrict_dev)
  File "/usr/lib/python3.2/site-packages/attic/archiver.py", line 169, in _process
    archive.process_item(path, st)
  File "/usr/lib/python3.2/site-packages/attic/archive.py", line 364, in process_item
    item.update(self.stat_attrs(st, path))
  File "/usr/lib/python3.2/site-packages/attic/archive.py", line 359, in stat_attrs
    acl_get(path, item, st, self.numeric_owner)
TypeError: acl_get() takes at most 3 positional arguments (4 given)
```
So I guess this issue of ACLs is more tricky and ingrained than I thought.
Anybody have any thoughts how to get this to work?
(PS: yes I know hard coding and bypassing things like I did doesn't mean
it will work that way...but I wanted to see how far I could get and what sort
of errors would rise up.)

**Comment by ThomasWaldmann** [comment]
In PR #235 there is a dummy implementation for xattr stuff now, so that part of the problem is solved, it will not just raise an exception any more for unsupported platforms, but just use the dummy xattr.

**Comment by ThomasWaldmann** [comment]
Fixing the ACL stuff now...

**Comment by jumper444** [comment]
Thank you. Will use and test soon.

**Comment by jumper444** [comment]
Issue corrected in your code changes above.

**Comment by ThomasWaldmann** [comment]
please reopen, it is not yet fixed in jborg/master.

**Comment by jumper444** [comment]
Reopening.  Sorry I'm a bit new with Git and the collaborative process.

**Comment by jumper444** [comment]
Pull request 235 (and the changes discussed herein) didn't make it into Attic 0.15 which was my expectation.  What is the process or criteria for this sort of thing? Thanks.

**Comment by jumper444** [comment]
One reason I ask is that these changes allow Attic on Cygwin platform which almost by definition is a less knowledgeable user (I'm including myself here. I installed Cygwin to get Attic...my default platform is windows).  It was rather involved for me to figure out how to pull and compile the patch version in order to use/test.  I'm not sure all people such as me would have the time (or background).  I think others who would otherwise be able to use and test  Cygwin Attice might have the same problem and I think having the xattr bypass patch in the PIP INSTALL version makes it more accessible to this group.  A warning print statement "Cygwin platform is untested" or something when run could alert people, but at least it would then open Attic up to even more users and "more eyeballs for bugs" as Linus says.

**Comment by jborg** [comment]
The now released Attic 0.15 is less strict about platform support for xattr and acl so other unsupported posix-like platforms like OpenBSD might now work out of the box. I'm not sure if Cygwin is posix-enough to work though, but I'm closing this issue now. Open new issues if you hit other problems.

That said, Windows, even with Cygwin is probably way to un-unix-like to ever work well with Attic. And I'm not interested in adding too much windows specific code to Attic.
